src/devices/Arduino/ArduinoBoard.cs

```diff
@@ -0,0 +1,514 @@
+        private Stream? _dataStream;
+        private FirmataDevice? _firmata;
+        private IReadOnlyList<SupportedPinConfiguration> _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+        private bool _initialized;
+        private bool _isDisposed = false;
+
+        // Counts how many spi devices are attached, to make sure we enable/disable the bus only when no devices are attached
+        private int _spiEnabled = 0;
+        private Version _firmwareVersion = new Version();
+        private string _firmwareName = string.Empty;
+        private Version _firmataVersion = new Version();
+
+        private object _initializationLock = new object();
+
+        /// <summary>
+        /// Creates an instance of an Ardino board connection using the given stream (typically from a serial port)
+        /// </summary>
+        /// <param name="serialPortStream">A stream to an Arduino/Firmata device</param>
+        public ArduinoBoard(Stream serialPortStream)
+        {
+            _dataStream = serialPortStream ?? throw new ArgumentNullException(nameof(serialPortStream));
+        }
+
+        /// <summary>
+        /// Creates an instance of the Arduino board connection connected to a serial port
+        /// </summary>
+        /// <param name="portName">Port name. On Windows, this is usually "COM3" or "COM4" for an Arduino attached via USB.
+        /// On Linux, possible values include "/dev/ttyAMA0", "/dev/serial0", "/dev/ttyUSB1", etc.</param>
+        /// <param name="baudRate">Baudrate to use. It is recommended to use at least 115200 Baud.</param>
+        public ArduinoBoard(string portName, int baudRate)
+        {
+            _dataStream = null;
+            _serialPort = new SerialPort(portName, baudRate);
+        }
+
+        /// <summary>
+        /// Attach to this event to retrieve log messages. The Exception argument may be null if it is only an informational message.
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
+
+        /// <summary>
+        /// Searches the given list of com ports for a firmata device.
+        /// </summary>
+        /// <remarks>
+        /// Scanning ports and testing for devices may affect unrelated devices. It is advisable to exclude ports known to contain other hardware from this scan.
+        /// A board won't be found if its port is already open (by the same or a different process).
+        /// </remarks>
+        /// <param name="comPorts">List of com ports. Can be used with <see cref="SerialPort.GetPortNames"/>.</param>
+        /// <param name="baudRates">List of baud rates to test. <see cref="CommonBaudRates"/>.</param>
+        /// <param name="board">[Out] Returns the board reference. It is already initialized.</param>
+        /// <returns>True on success, false if no board was found</returns>
+        public static bool TryFindBoard(IEnumerable<string> comPorts, IEnumerable<int> baudRates,
+#if !NETCOREAPP2_1
```

Comment:
@joperezr  can we do similar trick with attributes we do in dotnet/runtime so we can avoid `#if`s like this?